### ISSUE TITLE
Update the crates and generated code to rust 2018

### ISSUE THREAD

**Issue by Marwes** [lalrpop/lalrpop#515]
The rust code in the `.lalrpop` files are still on rust 2015 and need to be manually updated since `cargo fix` won't work on those files.

We should also update the generated code to emit 2018 idiomatic code (see #2018 )

**Comment by arashout** [comment]
I could work on this, but I don't really know enough about rust 2018 idioms, since I'm new to rust.
I couldn't find a list of "rust-2018-idioms" that I could use to go through and fix the issues. Got any tips?

**Comment by Marwes** [comment]
For the crates themselves you should be able to run `cargo fix`, `cargo fix --edition` and `cargo fix --edition-idioms` to get things up to date. The problem is that, with lalrpop's own parser being generated and stored in the repo we will also need to change the code that is generated to make it compatible.

The best way to go about this is probably to run each `cargo fix` command, check what changes it does to the generated parser, back out the changes and make code generation generate the updated code then run the fix command again (which should not modify the parser this time around).

There may be some chicken and egg issue when actually specifying `edition = "2018"` in the `Cargo.toml` but hopefully nothing to awkward.

**Comment by wangbyby** [comment]
I have done as ``cargo fix --edition`` tips me
```
To prepare for the 2018 edition you should first remove `edition = '2018'` from
your `Cargo.toml` and then rerun this command. Once all warnings have been fixed
then you can re-enable the `edition` key in `Cargo.toml`.
```
but can not fix the error

```
error[E0432]: unresolved import `ast`                                                           
 --> C:\Users\13\Desktop\lalrc\target\debug\build\lalrc-52069f57ba10be0e\out/calculator4.rs:4:5
  |
  | use ast::{Expr, Opcode};
  |     ^^^ help: a similar path exists: `crate::ast`
  |
  = note: `use` statements changed in Rust 2018; read more at <https://doc.rust-lang.org/edition-guide/rust-2018/module-system/path-clarity.html>
```
And the link is ``Not Found``. 
I also run the command ``cargo fix --edition-idioms``.

**Comment by wangbyby** [comment]
I fix my error by ``use crate::ast::{Expr, Opcode};``. 
So ``use crate::ast::{Expr, Opcode};`` can work.  
``use ast::{Expr, Opcode};`` just cause errors.

**Comment by yannham** [comment]
The crates have been updated to 2018 in #682. All the tests pass, but I'm intimate with the code generator enough to know if it generates 2018-compatible code (the tests passing at least tend to indicate it is the case, but they might be too simple to trigger issues). I propose to keep this issue open until we find out.

**Comment by Pat-Lafon** [comment]
Closing as resolved in #902
